### ISSUE TITLE
[Feature] Implement InternVL to llama.cpp

### ISSUE THREAD

**Issue by James4Ever0** [OpenGVLab/InternVL#522]
### Motivation

Many llama.cpp users are requesting this so far. Ollama is one of the interfaces of llama.cpp and it is quite popular. Implementing it will significantly accelerate InterVL adoption and recognition.

### Related resources

https://github.com/ggerganov/llama.cpp/issues/6803

### Additional context

InternVL is based on LLaMA architecture. Currently text-only InternLM models have been ported to Ollama but not for multimodal ones.

**Comment by G-z-w** [comment]
Thank you for your suggestions. We will gradually push support for various frameworks, and we also welcome contributions from the community.

**Comment by James4Ever0** [comment]
Hey @czczup would you please clarify the reason closing this issue?

**Comment by sammcj** [comment]
Any update on internVL support with llama.cpp?

**Comment by cloudyuyuyu** [comment]
Any update on internVL support with llama.cpp?

**Comment by G-z-w** [comment]
Thank you for your attention. We are actively progressing on this work, and we also welcome contributions from the community.

**Comment by rampageservices** [comment]
Just curious why this issue was closed if you are actively progressing on this work?

**Comment by rampageservices** [comment]
Thanks for reopening this issue.

**Comment by cloudyuyuyu** [comment]
any update?

**Comment by Cartomex-MX** [comment]
November 12, any update? thank in advance

**Comment by BleedingDev** [comment]
Is there anything that could be helped with? :) InternVL2.5 is really important for the future. :)

**Comment by linxhome** [comment]
pay close attention to this, any update? pls

**Comment by James4Ever0** [comment]
Is it really that hard to do this? I suggest myself to implement it.

**Comment by G-z-w** [comment]
> Is it really that hard to do this? I suggest myself to implement it.

Thank you for your willingness to help! We greatly appreciate your initiative and would be glad to have your contributions. If you need us to provide any content or information, feel free to let us know.

**Comment by James4Ever0** [comment]
Is [this model](https://huggingface.co/OpenGVLab/Mini-InternVL-Chat-2B-V1-5) based on the LLaVA architecture? What are the differences, in input, output and internal parameters?

**Comment by G-z-w** [comment]
> @G-z-w Is [this model](https://huggingface.co/OpenGVLab/Mini-InternVL-Chat-2B-V1-5) based on the LLaVA architecture? What are the differences, in input, output and internal parameters, and more?

This model, along with other InternVL chat models, is similar to the LLAVA framework, with the specific structure shown in the [link](https://huggingface.co/OpenGVLab/Mini-InternVL-Chat-2B-V1-5#model-details). The difference lies in dynamic resolution and pixel shuffle.  

If convenient, we recommend prioritizing the deployment of the [InternVL 2.5 series](https://huggingface.co/collections/OpenGVLab/internvl25-673e1019b66e2218f68d7c1c). The details of parameters are in [model card of blog](https://internvl.github.io/blog/2024-12-05-InternVL-2.5/).

**Comment by James4Ever0** [comment]
> > @G-z-w Is [this model](https://huggingface.co/OpenGVLab/Mini-InternVL-Chat-2B-V1-5) based on the LLaVA architecture? What are the differences, in input, output and internal parameters, and more?
> 
> This model, along with other InternVL chat models, is similar to the LLaVA framework, with the specific structure shown in the [link](https://huggingface.co/OpenGVLab/Mini-InternVL-Chat-2B-V1-5#model-details). The difference lies in dynamic resolution and pixel shuffle.  
> 
> If convenient, we recommend prioritizing the deployment of the [InternVL 2.5 series](https://huggingface.co/collections/OpenGVLab/internvl25-673e1019b66e2218f68d7c1c). The details of parameters are in [model card of blog](https://internvl.github.io/blog/2024-12-05-InternVL-2.5/).
> 

Is it identical to v1.5 series? Now I can run v1.5 on llama.cpp.

**Comment by G-z-w** [comment]
> > > @G-z-w Is [this model](https://huggingface.co/OpenGVLab/Mini-InternVL-Chat-2B-V1-5) based on the LLaVA architecture? What are the differences, in input, output and internal parameters, and more?
> > 
> > 
> > This model, along with other InternVL chat models, is similar to the LLaVA framework, with the specific structure shown in the [link](https://huggingface.co/OpenGVLab/Mini-InternVL-Chat-2B-V1-5#model-details). The difference lies in dynamic resolution and pixel shuffle.
> > If convenient, we recommend prioritizing the deployment of the [InternVL 2.5 series](https://huggingface.co/collections/OpenGVLab/internvl25-673e1019b66e2218f68d7c1c). The details of parameters are in [model card of blog](https://internvl.github.io/blog/2024-12-05-InternVL-2.5/).
> 
> Is the model structure of v2.5 identical to v1.5 series? Now I can run v1.5 on llama.cpp.
Yes, the structure of v2.5 is identical to that of the v1.5 series, except that the v2.5 series use different language models.

**Comment by BakingBrains** [comment]
@James4Ever0 any luck in running v2.5 model?

Thank you

**Comment by MrZeros** [comment]
any update?

**Comment by James4Ever0** [comment]
For anyone who abouts to work with the current code, you could check my latest release here.

It contains function-level diffs using universal ctags and some Python magic so anyone with entry-level C++ knowledge should be able to merge the changes easily.

**Comment by BrandonJull** [comment]
+1
